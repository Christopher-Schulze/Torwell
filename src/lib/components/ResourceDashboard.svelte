<script lang="ts">
  import { onMount } from 'svelte';
  import { listen } from '@tauri-apps/api/event';
  import MetricsChart from './MetricsChart.svelte';
  import type { MetricPoint } from '$lib/stores/torStore';

  const CHART_WIDTH = 120;
  const CHART_HEIGHT = 40;

  function buildPath(data: MetricPoint[], field: keyof MetricPoint): string {
    if (data.length === 0) return '';
    const maxVal = Math.max(...data.map((d) => d[field] as number), 1);
    const step = CHART_WIDTH / Math.max(data.length - 1, 1);
    let d = `M0 ${CHART_HEIGHT}`;
    data.forEach((pt, idx) => {
      const x = idx * step;
      const y = CHART_HEIGHT - ((pt[field] as number) / maxVal) * CHART_HEIGHT;
      d += ` L${x} ${y}`;
    });
    d += ` L${CHART_WIDTH} ${CHART_HEIGHT} Z`;
    return d;
  }

  $: cpuPath = buildPath(metrics, 'cpuPercent');
  $: networkPath = buildPath(metrics, 'networkBytes');

  let metrics: MetricPoint[] = [];
  const MAX_POINTS = 30;

  const MAX_MEMORY_MB = 1024;
  const MAX_CIRCUITS = 20;

  $: latest =
    metrics[metrics.length - 1] ?? {
      memoryMB: 0,
      circuitCount: 0,
      latencyMs: 0,
      oldestAge: 0,
      avgCreateMs: 0,
      failedAttempts: 0,
      cpuPercent: 0,
      networkBytes: 0,
      time: 0,
    };

  onMount(() => {
    const unlisten = listen<any>('metrics-update', (event) => {
      const point: MetricPoint = {
        time: Date.now(),
        memoryMB: Math.round(event.payload.memory_bytes / 1_000_000),
        circuitCount: event.payload.circuit_count,
        latencyMs: event.payload.latency_ms,
        oldestAge: event.payload.oldest_age ?? 0,
        avgCreateMs: event.payload.avg_create_ms ?? 0,
        failedAttempts: event.payload.failed_attempts ?? 0,
        cpuPercent: event.payload.cpu_percent ?? 0,
        networkBytes: event.payload.network_bytes ?? 0,
      };
      metrics = [...metrics, point].slice(-MAX_POINTS);
    });

    return () => {
      unlisten.then((fn) => fn());
    };
  });
</script>

<div class="glass-md rounded-xl p-4 space-y-4" role="region" aria-label="Resource dashboard">
  <div class="flex gap-4">
    <div class="flex-1">
      <p class="text-sm text-white">Memory: {latest.memoryMB} MB</p>
      {#if latest.memoryMB > MAX_MEMORY_MB}
        <p class="text-sm text-red-400" role="alert">Memory usage high</p>
      {/if}
    </div>
    <div class="flex-1">
      <p class="text-sm text-white">Circuits: {latest.circuitCount}</p>
      {#if latest.circuitCount > MAX_CIRCUITS}
        <p class="text-sm text-red-400" role="alert">Circuit count high</p>
      {/if}
    </div>
    <div class="flex-1">
      <p class="text-sm text-white">Avg build: {latest.avgCreateMs} ms</p>
    </div>
    <div class="flex-1">
      <p class="text-sm text-white">Failures: {latest.failedAttempts}</p>
    </div>
  </div>
  <div class="flex gap-4">
    <div class="flex-1">
      <p class="text-sm text-white">CPU: {latest.cpuPercent.toFixed(1)} %</p>
    </div>
    <div class="flex-1">
      <p class="text-sm text-white">Network: {latest.networkBytes} B/s</p>
    </div>
  </div>
  <div class="flex gap-2 items-end">
    <MetricsChart {metrics} />
    <svg
      width={CHART_WIDTH}
      height={CHART_HEIGHT}
      class="text-purple-400"
      role="img"
      aria-label="CPU usage chart"
    >
      {#if cpuPath}
        <path
          d={cpuPath}
          fill="currentColor"
          fill-opacity="0.3"
          stroke="currentColor"
          stroke-width="1"
        />
      {/if}
    </svg>
    <svg
      width={CHART_WIDTH}
      height={CHART_HEIGHT}
      class="text-cyan-400"
      role="img"
      aria-label="Network usage chart"
    >
      {#if networkPath}
        <path
          d={networkPath}
          fill="currentColor"
          fill-opacity="0.3"
          stroke="currentColor"
          stroke-width="1"
        />
      {/if}
    </svg>
  </div>
</div>

<style>
  .glass-md {
    @apply bg-white/20 backdrop-blur-xl;
  }
</style>
