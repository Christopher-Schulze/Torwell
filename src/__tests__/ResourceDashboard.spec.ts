import { render } from '@testing-library/svelte';
import { vi, describe, it, expect } from 'vitest';
import { tick } from 'svelte';

let metricsCallback: (e: any) => void = () => {};
vi.mock('@tauri-apps/api/event', () => ({
  listen: vi.fn((ev: string, cb: any) => {
    if (ev === 'metrics-update') metricsCallback = cb;
  })
}));

import ResourceDashboard from '../lib/components/ResourceDashboard.svelte';

describe('ResourceDashboard', () => {

  it('renders charts snapshot', async () => {
    const { container } = render(ResourceDashboard);
    await tick();

    metricsCallback({
      payload: {
        memory_bytes: 500_000_000,
        circuit_count: 5,
        latency_ms: 0,
        oldest_age: 0,
        avg_create_ms: 10,
        failed_attempts: 0,
        cpu_percent: 7.2,
      network_bytes: 1024,
      },
    });
    await tick();
    await tick();

    expect(container).toMatchSnapshot();
  });
});
