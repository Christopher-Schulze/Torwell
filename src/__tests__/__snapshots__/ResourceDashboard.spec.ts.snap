// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResourceDashboard > renders charts snapshot 1`] = `
<div>
  <div
    aria-label="Resource dashboard"
    class="glass-md rounded-xl p-4 space-y-4 svelte-umn5u1"
    role="region"
  >
    <div
      class="flex gap-4"
    >
      <div
        class="flex-1"
      >
        <p
          class="text-sm text-white"
        >
          Memory: 
          0
           MB
        </p>
         
      </div>
       
      <div
        class="flex-1"
      >
        <p
          class="text-sm text-white"
        >
          Circuits: 
          0
        </p>
         
      </div>
       
      <div
        class="flex-1"
      >
        <p
          class="text-sm text-white"
        >
          Avg build: 
          0
           ms
        </p>
      </div>
       
      <div
        class="flex-1"
      >
        <p
          class="text-sm text-white"
        >
          Failures: 
          0
        </p>
      </div>
    </div>
     
    <div
      class="flex gap-4"
    >
      <div
        class="flex-1"
      >
        <p
          class="text-sm text-white"
        >
          CPU: 
          0.0
           %
        </p>
      </div>
       
      <div
        class="flex-1"
      >
        <p
          class="text-sm text-white"
        >
          Network: 
          0
           B/s
        </p>
      </div>
    </div>
     
    <div
      class="flex gap-2 items-end"
    >
      <svg
        aria-label="Tor metrics chart"
        class="text-green-400"
        height="40"
        role="img"
        width="120"
      >
        
        
      </svg>
       
      <svg
        aria-label="CPU usage chart"
        class="text-purple-400"
        height="40"
        role="img"
        width="120"
      />
       
      <svg
        aria-label="Network usage chart"
        class="text-cyan-400"
        height="40"
        role="img"
        width="120"
      />
    </div>
  </div>
</div>
`;
